File: ios/Packages/AppFeatures/Sources/LeaderboardFeature/LeaderboardModels.swift
import Foundation

public struct LeaderboardEntry: Decodable, Equatable, Sendable {
    public let rank: Int
    public let score: Double
    public let displayName: String

    enum CodingKeys: String, CodingKey {
        case rank, score
        case displayName = "display_name"
    }
}

public struct LeaderboardBoard: Decodable, Equatable, Identifiable, Sendable {
    public var id: String { key }
    public let key: String
    public let name: String
    public let scope: String
    public let period: String
    public let joined: Bool
    public let myRank: Int?
    public let entries: [LeaderboardEntry]
    /// The current period's calendar bounds, for a "resets on" display — not
    /// the scoring window (which ends at "now", not the boundary). `nil` for
    /// `lifetime` (no reset) and for `seasonal`/`challenge` boards an admin
    /// hasn't dated yet.
    public let periodStartsAt: Date?
    public let periodEndsAt: Date?

    public init(
        key: String, name: String, scope: String, period: String,
        joined: Bool, myRank: Int?, entries: [LeaderboardEntry],
        periodStartsAt: Date? = nil, periodEndsAt: Date? = nil
    ) {
        self.key = key
        self.name = name
        self.scope = scope
        self.period = period
        self.joined = joined
        self.myRank = myRank
        self.entries = entries
        self.periodStartsAt = periodStartsAt
        self.periodEndsAt = periodEndsAt
    }

    enum CodingKeys: String, CodingKey {
        case key, name, scope, period, joined, entries
        case myRank = "my_rank"
        case periodStartsAt = "period_starts_at"
        case periodEndsAt = "period_ends_at"
    }
}

public struct LeaderboardMembership: Codable, Equatable, Sendable {
    public let handle: String
    public let publishName: Bool
    public let city: String?

    enum CodingKeys: String, CodingKey {
        case handle
        case publishName = "publish_name"
        case city
    }
}

struct ListBoardsResponse: Decodable {
    let boards: [LeaderboardBoard]
}

struct JoinLeaderboardRequest: Encodable {
    let publish_name: Bool
    let city: String?
    /// ISO 3166-1 alpha-2. Sent only for country-scope boards; the backend
    /// stores it only for those, and rejects the join without it.
    let country: String?
}

// `city` clearing (explicit null) is a PATCH-only edge case not exposed by the
// join/leave-driven UI in this pass — leaving a board already clears its city
// association (docs/features/leaderboard.md), so only additive updates are needed here.
struct UpdateMembershipRequest: Encodable {
    let publish_name: Bool?
    let city: String?
}

File: android/feature/leaderboard/src/main/kotlin/com/fatwabot/feature/leaderboard/LeaderboardModels.kt
package com.fatwabot.feature.leaderboard

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class LeaderboardEntry(
    val rank: Int,
    val score: Double,
    @SerialName("display_name") val displayName: String,
)

@Serializable
data class LeaderboardBoard(
    val key: String,
    val name: String,
    val scope: String,
    val period: String,
    val joined: Boolean,
    @SerialName("my_rank") val myRank: Int? = null,
    val entries: List<LeaderboardEntry> = emptyList(),
    /**
     * The current period's calendar bounds as raw ISO-8601 strings (parsed at
     * the display layer, same convention as [com.fatwabot.core.common.QueuedAnalyticsEvent]'s
     * `occurredAt`) — for a "resets on" display, not the scoring window (which
     * ends at "now", not the boundary). `null` for `lifetime` (no reset) and
     * for a `seasonal`/`challenge` board an admin hasn't dated yet.
     */
    @SerialName("period_starts_at") val periodStartsAt: String? = null,
    @SerialName("period_ends_at") val periodEndsAt: String? = null,
)

@Serializable
data class LeaderboardMembership(
    val handle: String,
    @SerialName("publish_name") val publishName: Boolean,
    val city: String? = null,
    val country: String? = null,
)

@Serializable
internal data class ListBoardsResponse(val boards: List<LeaderboardBoard>)

@Serializable
internal data class JoinLeaderboardRequest(
    @SerialName("publish_name") val publishName: Boolean,
    val city: String? = null,
    /**
     * ISO 3166-1 alpha-2. Sent only for country-scope boards; the backend
     * stores it only for those, and rejects the join without it.
     */
    val country: String? = null,
)

@Serializable
internal data class LeftResponse(val left: Boolean)
